File: Chrome-Ext-Capture-API/src/popupStats.js
// Pure popup recording stats builder — mirrors Go package popupstats rules.
// Used by background getState enrichment (no Chrome APIs inside).

const OPAQUE_HOST = 'opaque';

/**
 * Parse hostname from request URL; invalid/empty → "opaque".
 * @param {string} raw
 * @returns {string}
 */
function hostFromURL(raw) {
  if (!raw) return OPAQUE_HOST;
  try {
    // Relative / scheme-less strings: URL() may throw or yield empty host.
    const u = new URL(raw);
    const host = u.hostname;
    if (!host) return OPAQUE_HOST;
    return host;
  } catch (_) {
    return OPAQUE_HOST;
  }
}

/**
 * Preferred entry key: `${tabId}:${requestId}` — tabId before first ':'.
 * @param {string} key
 * @returns {number|null}
 */
function tabIDFromKey(key) {
  if (typeof key !== 'string') return null;
  const idx = key.indexOf(':');
  if (idx <= 0) return null;
  const n = Number.parseInt(key.slice(0, idx), 10);
  if (!Number.isFinite(n) || String(n) !== key.slice(0, idx)) return null;
  return n;
}

/**
 * Top-N domains by count desc, host asc on ties.
 * @param {Map<string, number>|Object<string, number>} hosts
 * @param {number} n
 * @returns {{host: string, count: number}[]}
 */
function topDomains(hosts, n) {
  const list = [];
  if (hosts instanceof Map) {
    for (const [host, count] of hosts) list.push({ host, count });
  } else {
    for (const host of Object.keys(hosts || {})) {
      list.push({ host, count: hosts[host] });
    }
  }
  list.sort((a, b) => {
    if (a.count !== b.count) return b.count - a.count;
    return a.host < b.host ? -1 : a.host > b.host ? 1 : 0;
  });
  if (n > 0 && list.length > n) return list.slice(0, n);
  return list;
}

/**
 * Build enriched popup stats from in-memory capture state.
 *
 * @param {{
 *   entries: Object<string, {request?: {url?: string}, url?: string}|string>,
 *   attachedTabIds: Iterable<number>|number[],
 *   tabMeta?: Object<number, {title?: string, url?: string, active?: boolean}>
 * }} input
 * @returns {{
 *   count: number,
 *   tabsWatching: number,
 *   domainCount: number,
 *   tabs: Array<{
 *     tabId: number,
 *     title: string,
 *     url: string,
 *     active: boolean,
 *     attached: boolean,
 *     requestCount: number,
 *     domainCount: number,
 *     domains: {host: string, count: number}[]
 *   }>
 * }}
 */
export function buildPopupStats(input) {
  const entries = input && input.entries ? input.entries : {};
  const tabMeta = (input && input.tabMeta) || {};
  const attachedSet = new Set();
  const attachedList = input && input.attachedTabIds != null
    ? Array.from(input.attachedTabIds)
    : [];
  for (const id of attachedList) {
    if (id != null && Number.isFinite(Number(id))) attachedSet.add(Number(id));
  }

  /** @type {Map<number, {hosts: Map<string, number>, reqCount: number}>} */
  const byTab = new Map();
  const globalHosts = new Set();

  const entryKeys = Object.keys(entries);
  for (const key of entryKeys) {
    const raw = entries[key];
    let url = '';
    if (typeof raw === 'string') {
      url = raw;
    } else if (raw && typeof raw === 'object') {
      url = (raw.request && raw.request.url) || raw.url || '';
    }
    const host = hostFromURL(url);
    globalHosts.add(host);

    const tabId = tabIDFromKey(key);
    // Fallback: HAR-style _tabId on entry object.
    let tid = tabId;
    if (tid == null && raw && typeof raw === 'object' && raw._tabId != null) {
      const n = Number(raw._tabId);
      if (Number.isFinite(n)) tid = n;
    }
    if (tid == null) continue;

    let acc = byTab.get(tid);
    if (!acc) {
      acc = { hosts: new Map(), reqCount: 0 };
      byTab.set(tid, acc);
    }
    acc.reqCount += 1;
    acc.hosts.set(host, (acc.hosts.get(host) || 0) + 1);
  }

  const memberIds = new Set(attachedSet);
  for (const id of byTab.keys()) memberIds.add(id);

  const tabs = [];
  for (const id of memberIds) {
    const attached = attachedSet.has(id);
    const meta = tabMeta[id] || tabMeta[String(id)] || {};
    const acc = byTab.get(id);
    const reqCount = acc ? acc.reqCount : 0;
    const hosts = acc ? acc.hosts : new Map();
    const domainCount = hosts.size;
    const domains = topDomains(hosts, 3);

    let title = meta.title || '';
    if (!title) {
      title = attached ? `Tab ${id}` : 'Closed tab';
    }

    tabs.push({
      tabId: id,
      title,
      url: meta.url || '',
      active: !!meta.active,
      attached,
      requestCount: reqCount,
      domainCount,
      domains,
    });
  }

  tabs.sort((a, b) => {
    if (a.requestCount !== b.requestCount) return b.requestCount - a.requestCount;
    return a.tabId - b.tabId;
  });

  return {
    count: entryKeys.length,
    tabsWatching: attachedSet.size,
    domainCount: globalHosts.size,
    tabs,
  };
}

export { OPAQUE_HOST, hostFromURL, tabIDFromKey, topDomains };

File: popupstats/popupstats.go
// Package popupstats builds pure popup recording stats for API Capture.
// Rules mirror Chrome-Ext-Capture-API getState enrichment (see popupStats.js).
package popupstats

import (
	"fmt"
	"net/url"
	"sort"
	"strconv"
	"strings"
)

// OpaqueHost is the single bucket name for invalid/unparseable request URLs.
const OpaqueHost = "opaque"

// EntryValue is one capture buffer entry (request.url only for stats).
type EntryValue struct {
	URL string // request.url
}

// TabMeta is chrome.tabs-style metadata for a tab id.
type TabMeta struct {
	Title  string
	URL    string
	Active bool
}

// DomainCount is a host and how many requests hit it.
type DomainCount struct {
	Host  string
	Count int
}

// TabStats is one row in the popup "By tab" list.
type TabStats struct {
	TabID        int
	Title        string
	URL          string
	Active       bool
	Attached     bool
	RequestCount int
	DomainCount  int
	Domains      []DomainCount // top 3 only
}

// PopupStats is the enriched getState payload shape for recording chips + list.
type PopupStats struct {
	Count        int
	TabsWatching int
	DomainCount  int // unique hosts across all entries
	Tabs         []TabStats
}

// Input is the pure builder input (in-memory capture state).
type Input struct {
	Entries        map[string]EntryValue // key "${tabId}:${requestId}" preferred
	AttachedTabIDs []int
	TabMeta        map[int]TabMeta
}

type tabAccum struct {
	tabID    int
	hosts    map[string]int
	reqCount int
}

// BuildPopupStats turns capture buffer + attached set + tab meta into popup stats.
func BuildPopupStats(in Input) PopupStats {
	entries := in.Entries
	if entries == nil {
		entries = map[string]EntryValue{}
	}
	tabMeta := in.TabMeta
	if tabMeta == nil {
		tabMeta = map[int]TabMeta{}
	}

	attachedSet := make(map[int]struct{}, len(in.AttachedTabIDs))
	for _, id := range in.AttachedTabIDs {
		attachedSet[id] = struct{}{}
	}

	// Per-tab host counts + membership from entries.
	byTab := make(map[int]*tabAccum)
	globalHosts := make(map[string]struct{})

	for key, ev := range entries {
		tabID, ok := tabIDFromKey(key)
		if !ok {
			// Still count globally so chips reflect buffer size; skip per-tab
			// attribution for unparseable keys (production always uses tabId:...).
			host := hostFromURL(ev.URL)
			globalHosts[host] = struct{}{}
			continue
		}
		acc := byTab[tabID]
		if acc == nil {
			acc = &tabAccum{tabID: tabID, hosts: make(map[string]int)}
			byTab[tabID] = acc
		}
		acc.reqCount++
		host := hostFromURL(ev.URL)
		acc.hosts[host]++
		globalHosts[host] = struct{}{}
	}

	// Union: attached OR has ≥1 entry.
	memberIDs := make(map[int]struct{}, len(attachedSet)+len(byTab))
	for id := range attachedSet {
		memberIDs[id] = struct{}{}
	}
	for id := range byTab {
		memberIDs[id] = struct{}{}
	}

	tabs := make([]TabStats, 0, len(memberIDs))
	for id := range memberIDs {
		_, attached := attachedSet[id]
		meta := tabMeta[id]
		acc := byTab[id]

		reqCount := 0
		var hosts map[string]int
		if acc != nil {
			reqCount = acc.reqCount
			hosts = acc.hosts
		}
		if hosts == nil {
			hosts = map[string]int{}
		}

		domainCount := len(hosts)
		domains := topDomains(hosts, 3)

		title := meta.Title
		if title == "" {
			if attached {
				title = fmt.Sprintf("Tab %d", id)
			} else {
				title = "Closed tab"
			}
		}

		tabs = append(tabs, TabStats{
			TabID:        id,
			Title:        title,
			URL:          meta.URL,
			Active:       meta.Active,
			Attached:     attached,
			RequestCount: reqCount,
			DomainCount:  domainCount,
			Domains:      domains,
		})
	}

	sort.SliceStable(tabs, func(i, j int) bool {
		if tabs[i].RequestCount != tabs[j].RequestCount {
			return tabs[i].RequestCount > tabs[j].RequestCount
		}
		return tabs[i].TabID < tabs[j].TabID
	})

	// Entries that failed tab attribution still count toward global count.
	count := len(entries)

	return PopupStats{
		Count:        count,
		TabsWatching: len(attachedSet),
		DomainCount:  len(globalHosts),
		Tabs:         tabs,
	}
}

// tabIDFromKey parses the integer tab id before the first ':' in entry key.
func tabIDFromKey(key string) (int, bool) {
	idx := strings.IndexByte(key, ':')
	if idx <= 0 {
		return 0, false
	}
	id, err := strconv.Atoi(key[:idx])
	if err != nil {
		return 0, false
	}
	return id, true
}

// hostFromURL returns the URL hostname, or OpaqueHost when unparseable/empty.
func hostFromURL(raw string) string {
	if raw == "" {
		return OpaqueHost
	}
	u, err := url.Parse(raw)
	if err != nil {
		return OpaqueHost
	}
	host := u.Hostname()
	if host == "" {
		return OpaqueHost
	}
	return host
}

// topDomains returns up to n hosts sorted by count desc, then host asc.
func topDomains(hosts map[string]int, n int) []DomainCount {
	if len(hosts) == 0 {
		return nil
	}
	list := make([]DomainCount, 0, len(hosts))
	for h, c := range hosts {
		list = append(list, DomainCount{Host: h, Count: c})
	}
	sort.SliceStable(list, func(i, j int) bool {
		if list[i].Count != list[j].Count {
			return list[i].Count > list[j].Count
		}
		return list[i].Host < list[j].Host
	})
	if n > 0 && len(list) > n {
		list = list[:n]
	}
	return list
}
